src/workerd/api/sockets.h

```diff
@@ -103,7 +127,31 @@ class Socket: public jsg::Object {
   jsg::Ref<ReadableStream> readable;
   jsg::Ref<WritableStream> writable;
   kj::Promise<kj::Own<kj::AsyncIoStream>> processConnection();
-  IoOwn<kj::PromiseFulfillerPair<void>> closeFulfiller;
+  IoOwn<kj::PromiseFulfillerPair<bool>> closeFulfiller;
+  // This fulfiller is used to signal either an implicit or explicit socket closure.
+  IoOwn<kj::PromiseFulfillerPair<void>> jsCloseFulfiller;
+  // Used to signal to JS scripts that socket has been closed. Its `promise` is returned via the
+  // `closed` property of the socket.
+
+  kj::ForkedPromise<void> jsCloseFulfillerFork;
+  void performClose(bool isImplicit);
+
+  void resolveJsFulfiller(kj::Maybe<kj::Exception> maybeErr) {
```

Comment:
Instead of using a KJ promise and fulfiller for `jsCloseFulfiller`, perhaps it would be better to use `jsg::Lock::newPromiseAndResolver()` to create a JavaScript promise/resolver pair? That way, the event is signaled through JavaScript microtasks rather than the KJ event loop, which avoids the need to exit and re-enter the isolate. You can also return the same promise object each time `getClosed()` is called (use `jsg::MemoizedIdentity` for this).

Don't forget to implement GC visitation for these objects.